### PR TITLE
Add "feat. Artist" filter for Amazon tracks, Album Artist data from Artist tag and improved Explicit tag handling

### PR DESCRIPTION
**Pull request by clasick**
**Describe the changes you made**
Added a filter that removes "[feat. Artist]" tags from the track's `Title` metadata in Amazon Music.

Some tracks contain "(Explicit Version)" that needed to be handling for "Explicit" filtering. Added this as a case under "Version" filtering. 

**Additional context**
Effectively, all the changes combined turn this:
image
into this:
image

### CODE REVIEW

**Comment by alexesprit** [comment]
Does Amazon always contain featuring artists in "Artist" field? I believe the common format is `Artist A — Track (feat. Artist B)`, but not `Artist A feat. Artist B — Track`.

**Comment by clasick** [comment]
> Does Amazon always contain featuring artists in "Artist" field?
> 
> I believe the common format is `Artist A — Track (feat. Artist B)`, but not `Artist A feat. Artist B — Track`. What about removing featuring artists from `artist` field?

Yes, I looked into many albums and they always contain the format `Artist A feat. Artist B — Track`. I was going to fix this next, except I had a doubt. By most conventions on other scrobblers I've seen, they keep the `Artist` tag as how it's present in Amazon. The `Album Artist` tag would be the one withe the `feat. Artist B` part removed. 

How is it generally implemented on `web-scrobbler`? Following conventions, we can keep the `Artist` as is and add data for `Album Artist` and send both in the scrobble data.

**Comment by alexesprit** [comment]
> Yes, I looked into many albums and they always contain the format Artist A feat. Artist B — Track. I was going to fix this next, except I had a doubt. By most conventions on other scrobblers I've seen, they keep the Artist tag as how it's present in Amazon.

I would keep the current behavior in PR, if `Artist` field **always** contains featuring artists, and `Track` field contains featuring artists **sometimes**.

Otherwise, I would remove featuring artists from `Track` field.

> How is it generally implemented on web-scrobbler?

In most cases, connectors return track info as it's displayed on website; some connectors implement their own filters to modify track fields. The core does not modify this info directly before scrobbling; it sends a request to Last.fm to correct track info, though).

**Comment by alexesprit** [comment]
Could you also rebase to latest master and add tests for new filter rules? See [tests/content/filter.js][tests] for examples.

[tests]: https://github.com/web-scrobbler/web-scrobbler/blob/master/tests/content/filter.js

**Comment by clasick** [comment]
> > Yes, I looked into many albums and they always contain the format Artist A feat. Artist B — Track. I was going to fix this next, except I had a doubt. By most conventions on other scrobblers I've seen, they keep the Artist tag as how it's present in Amazon.
> 
> I would keep the current behavior in PR, if `Artist` field **always** contains featuring artists, and `Track` field contains featuring artists **sometimes**.

Yes, this is the case. I was thinking scrobbling "Album Artist" field in this case with "feat. Artist" removed along with "Artist" would be the correct move, and let Last.fm figure it out on their end. (please refer to the answer to my issue in another project [here](https://github.com/hrkfdn/mpdas/issues/48#issuecomment-341999852)).

ex. `Artist` = The Weeknd feat. Ed Sheeran
      `Album Artist` = The Weeknd

> Otherwise, I would remove featuring artists from `Track` field.
Yes, it is how the code in this PR works right now.

> > How is it generally implemented on web-scrobbler?
> 
> In most cases, connectors return track info as it's displayed on website; some connectors implement their own filters to modify track fields. The core does not modify this info directly before scrobbling; it sends a request to Last.fm to correct track info, though).

I was thinking we should formulate some kind of standard similar to what Last.fm themselves suggested in their [Correction guidelines](https://getsatisfaction.com/lastfm/topics/what-are-the-guidelines-for-correcting-artist-and-track-names) and write connectors which adher to that, because the user would be expecting scrobbling data to be handled in a similar way throughout, and not differing between sites.

> Could you also rebase to latest master and add tests for new filter rules? See tests/content/filter.js for examples.

Sure, I'll get on this.

**Comment by clasick** [comment]
Another idea is to scrobble both by default (for all connectors) and let Last.fm decide what to do with it. We can then add a option toggle in the Preferences that can force `Album Artist` to be scrobbled as `Artist` if the user desires (many users including myself scrobble it this way).

**Comment by clasick** [comment]
I can confirm this behavior now. Scrobbling with both `Artist` (with feat. Artist B) and `Album Artist` (without feat. Artist B) leads Last.fm to intelligently use it for both formats i.e. the scrobble will show up under the `Artist A` while also under `Artist A feat. B` while counting as only one unique scrobble.

File: src/connectors/amazon.js [hunk 1]
```diff
@@ -6,9 +6,17 @@ Connector.getArtist = () => {
...
+
 Connector.trackSelector = '.trackInfoContainer .trackTitle';
 
+Connector.albumSelector = '.trackSourceLink a';
+
 Connector.getAlbum = () => {
+	if ($('.trackSourceLink a').attr('href').includes('albums')) {
+		return $('.trackSourceLink a').text();
+	}
+
```
Comment (clasick) [inline src/connectors/amazon.js]: Album tag was broken. This check fixes that.

**Comment by clasick** [comment]
Looks like I broke some tests because of function renaming. Will fix.

**Comment by clasick** [comment]
Album Artist from Artist in action:
image
gets scrobbled as:
image

**Comment by alexesprit** [comment]
Awesome work. I have no access to Amazon, so I can't verify changes. I believe you done things right.

File: src/connectors/amazon.js (hunk 1 above)
Comment (alexesprit) [reply src/connectors/amazon.js]: Does this change fix #1677?

**Comment by clasick** [comment]
> Awesome work. I have no access to Amazon, so I can't verify changes. I believe you done things right.

Yup, I've been using this test version for the past couple days, seems to be working fine.

There are a couple of other issues in Amazon that happen rarely, I'll capture try to address those as well later on.

**Pull request merged**